File: Caronae/Ride/RideListController.m
#import <CRToast/CRToast.h>
#import "CaronaeAlertController.h"
#import "RideListController.h"
#import "UIViewController+isVisible.h"
#import "Caronae-Swift.h"

static NSString *const RideListDefaultEmptyMessage = @"Nenhuma carona\nencontrada.";
static NSString *const RideListDefaultLoadingMessage = @"Carregando...";
static NSString *const RideListDefaultErrorMessage = @"Não foi possível\ncarregar as caronas.";

static CGFloat const RideListMessageFontSize = 25.0f;

@interface RideListController()
@property (nonatomic, readwrite) NSArray<Ride *> *filteredRides;
@end

@implementation RideListController

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.view = [[[NSBundle mainBundle] loadNibNamed:NSStringFromClass(RideListController.class)
                                                   owner:self
                                                 options:nil] objectAtIndex:0];
        
        self.historyTable = NO;
        self.ridesDirectionGoing = YES;
        self.hidesDirectionControl = NO;
        
        if ([self respondsToSelector:NSSelectorFromString(@"refreshTable")]) {
            self.refreshControl = [[UIRefreshControl alloc] init];
            self.refreshControl.tintColor = [UIColor colorWithWhite:0.9f alpha:1.0f];
            [self.refreshControl addTarget:self
                                    action:NSSelectorFromString(@"refreshTable")
                          forControlEvents:UIControlEventValueChanged];
            [self.tableView addSubview:self.refreshControl];
        }
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UINib *cellNib = [UINib nibWithNibName:NSStringFromClass(RideCell.class) bundle:nil];
    [self.tableView registerNib:cellNib forCellReuseIdentifier:@"Ride Cell"];
    
    if (self.hidesDirectionControl) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.directionControl removeFromSuperview];
        });
    }
    else {
        self.tableView.contentInset = UIEdgeInsetsMake(45.0f, 0.0f, 0.0f, 0.0f);
    }
    
    self.tableView.rowHeight = 85.0f;
    self.tableView.backgroundView = self.loadingLabel;
    
    if (self.historyTable) {
        self.tableView.allowsSelection = NO;
    }
}

- (void)setRides:(NSArray *)rides {
    _rides = rides;
    [self updateFilteredRides];
}

- (void)updateFilteredRides {
    if (_rides) {
        _tableView.backgroundView = ([_rides count] == 0) ? self.emptyTableLabel : nil;
        
        NSMutableArray *filtered = [NSMutableArray arrayWithCapacity:[_rides count]];
        for (Ride *ride in _rides) {
            if (self.hidesDirectionControl || ride.going == self.ridesDirectionGoing) {
                [filtered addObject:ride];
            }
        }
        
        self.filteredRides = filtered;
    } else {
        self.filteredRides = @[];
    }
}

- (void)loadingFailedWithError:(NSError *)error {
    if (self.filteredRides.count == 0) {
        self.tableView.backgroundView = self.errorLabel;
    }
    
    NSLog(@"%@ failed to load rides: %@", NSStringFromClass(self.class), error.localizedDescription);
    
    if (![self isVisible]) return;
    
    if ([error.domain isEqualToString:NSURLErrorDomain]) {
        if (error.code == NSURLErrorNotConnectedToInternet) {
            [CRToastManager showNotificationWithOptions:@{
                                                          kCRToastTextKey: @"Sem conexão com a internet",
                                                          kCRToastBackgroundColorKey: [UIColor redColor],
                                                          }
                                        completionBlock:nil];
            return;
        }
        
        if (error.code == NSURLErrorTimedOut ||
            error.code == NSURLErrorCannotFindHost ||
            error.code == NSURLErrorCannotConnectToHost ||
            error.code == NSURLErrorNetworkConnectionLost) {
            [CRToastManager showNotificationWithOptions:@{
                                                          kCRToastTextKey: @"Sem conexão com o Caronaê",
                                                          kCRToastBackgroundColorKey: [UIColor redColor],
                                                          }
                                        completionBlock:nil];
            return;
        }
    }
    
    NSString *errorAlertTitle = @"Algo deu errado";
    NSString *errorAlertMessage = [NSString stringWithFormat:@"Não foi possível carregar as caronas. Por favor, tente novamente. (%@)", error.localizedDescription];
    [CaronaeAlertController presentOkAlertWithTitle:errorAlertTitle message:errorAlertMessage];
}


#pragma mark - Table methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.filteredRides count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    RideCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Ride Cell" forIndexPath:indexPath];
    
    if (!self.historyTable) {
        [cell configureCellWithRide:self.filteredRides[indexPath.row]];
    }
    else {
        [cell configureHistoryCellWithRide:self.filteredRides[indexPath.row]];
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if (!self.historyTable) {
        Ride *ride = self.filteredRides[indexPath.row];
        RideViewController *rideVC = [RideViewController rideViewControllerForRide:ride];
        
        [self.navigationController pushViewController:rideVC animated:YES];
    }
}


#pragma mark - IBActions

- (IBAction)didChangeDirection:(UISegmentedControl *)sender {
    self.ridesDirectionGoing = (sender.selectedSegmentIndex == 0);
    [self setRides:self.rides];
    [self.tableView reloadData];
}


#pragma mark - Extra views

// Background view when the table is empty
- (UILabel *)emptyTableLabel {
    if (!_emptyTableLabel) {
        _emptyTableLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
        _emptyTableLabel.text = self.emptyMessage ? self.emptyMessage : RideListDefaultEmptyMessage;
        _emptyTableLabel.textColor = [UIColor grayColor];
        _emptyTableLabel.numberOfLines = 0;
        _emptyTableLabel.textAlignment = NSTextAlignmentCenter;
        _emptyTableLabel.font = [UIFont systemFontOfSize:RideListMessageFontSize weight:UIFontWeightUltraLight];
        [_emptyTableLabel sizeToFit];
    }
    return _emptyTableLabel;
}

// Background view when an error occurs
- (UILabel *)errorLabel {
    if (!_errorLabel) {
        _errorLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
        _errorLabel.text = RideListDefaultErrorMessage;
        _errorLabel.textColor = [UIColor grayColor];
        _errorLabel.numberOfLines = 0;
        _errorLabel.textAlignment = NSTextAlignmentCenter;
        _errorLabel.font = [UIFont systemFontOfSize:RideListMessageFontSize weight:UIFontWeightUltraLight];
        [_errorLabel sizeToFit];
    }
    return _errorLabel;
}

// Background view when the table is loading
- (UILabel *)loadingLabel {
    if (!_loadingLabel) {
        _loadingLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
        _loadingLabel.text = RideListDefaultLoadingMessage;
        _loadingLabel.textColor = [UIColor grayColor];
        _loadingLabel.numberOfLines = 0;
        _loadingLabel.textAlignment = NSTextAlignmentCenter;
        _loadingLabel.font = [UIFont systemFontOfSize:RideListMessageFontSize weight:UIFontWeightUltraLight];
        [_loadingLabel sizeToFit];
    }
    return _loadingLabel;
}

@end

File: Caronae/Ride/RideListController.swift
import UIKit
import RealmSwift

class RideListController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var directionControl: UISegmentedControl!
    @IBOutlet weak var filterLabel: UILabel!
    @IBOutlet weak var filterView: UIView!
    @IBOutlet weak var filterViewHeight: NSLayoutConstraint!
    @IBOutlet weak var filterViewHeightZero: NSLayoutConstraint!
    @IBInspectable var emptyMessage: String?
    @IBInspectable var historyTable: Bool = false
    
    struct RideListDefault {
        static let emptyMessage = "Nenhuma carona\nencontrada."
        static let loadingMessage = "Carregando..."
        static let errorMessage = "Não foi possível\ncarregar as caronas."
        
        static let messageFontSize: CGFloat = 25.0
    }
    
    var ridesDirectionGoing = true
    var hidesDirectionControl = false
    var filterIsEnabled = false
    
    var refreshControl: UIRefreshControl?

    var rides = [Ride]() {
        didSet {
            self.updateFilteredRides()
        }
    }
    var filteredRides = [Ride]()
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        self.view = Bundle.main.loadNibNamed(String(describing: RideListController.self), owner: self, options: nil)?.first as! UIView
        
        if self.responds(to: #selector(refreshTable)) {
            refreshControl = UIRefreshControl()
            refreshControl?.tintColor = UIColor(white: 0.9, alpha: 1.0)
            refreshControl?.addTarget(self, action: #selector(refreshTable), for: .valueChanged)
            self.tableView.addSubview(refreshControl!)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if #available(iOS 11.0, *) {
            // Create constraint between filterView and safe area layout guide
            let safeArea = self.view.safeAreaLayoutGuide
            filterView.topAnchor.constraint(equalTo: safeArea.topAnchor).isActive = true
        }
        
        tableView.delegate = self
        tableView.dataSource = self
        let cellNib = UINib(nibName: String(describing: RideCell.self), bundle: nil)
        tableView.register(cellNib, forCellReuseIdentifier: "Ride Cell")
        
        if hidesDirectionControl {
            DispatchQueue.main.async {
                self.directionControl.removeFromSuperview()
            }
        } else {
            // Configure direction titles according to institution
            directionControl.setTitle(PlaceService.Institution.goingLabel, forSegmentAt: 0)
            directionControl.setTitle(PlaceService.Institution.leavingLabel, forSegmentAt: 1)
        }
        adjustTableView()
        
        tableView.rowHeight = 85.0
        tableView.backgroundView = loadingLabel
        
        if historyTable {
            tableView.allowsSelection = false
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        adjustFilterView(animated: false)
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        adjustTableView()
    }
    
    @objc func refreshTable() {
    }
    
    func adjustFilterView(animated: Bool) {
        if hidesDirectionControl {
            view.layoutIfNeeded()
            filterViewHeightZero.isActive = true
            return
        }
        
        if filterIsEnabled {
            view.layoutIfNeeded()
            filterViewHeightZero.isActive = false
            filterView.layer.masksToBounds = false
        } else {
            filterView.layer.masksToBounds = true
            if animated {
                filterViewHeightZero.isActive = true
                UIView.animate(withDuration: 0.2, animations: {
                    self.view.layoutIfNeeded()
                })
            } else {
                view.layoutIfNeeded()
                filterViewHeightZero.isActive = true
            }
        }
    }
    
    func adjustTableView() {
        guard !hidesDirectionControl else {
            return
        }
        
        var defaultTopInset: CGFloat = 64.0
        var defaultBottomInset: CGFloat = 49.0
        if #available(iOS 11.0, *) {
            defaultTopInset = 0.0
            defaultBottomInset = 0.0
        }
        
        guard let navigationBarHeight = self.navigationController?.navigationBar.frame.size.height else {
            return
        }
        if filterIsEnabled {
            let filterViewHeight = self.filterViewHeight.constant
            tableView.scrollIndicatorInsets = UIEdgeInsets(top: defaultTopInset + filterViewHeight, left: 0, bottom: defaultBottomInset, right: 0)
            tableView.contentInset = UIEdgeInsets(top: defaultTopInset + navigationBarHeight + filterViewHeight, left: 0, bottom: defaultBottomInset, right: 0)
        } else {
            tableView.scrollIndicatorInsets = UIEdgeInsets(top: defaultTopInset, left: 0, bottom: defaultBottomInset, right: 0)
            tableView.contentInset = UIEdgeInsets(top: defaultTopInset + navigationBarHeight, left: 0, bottom: defaultBottomInset, right: 0)
        }
    }
    
    func updateFilteredRides() {
        tableView.backgroundView = rides.isEmpty ? emptyTableLabel : nil
        
        if hidesDirectionControl {
            filteredRides = rides
        } else {
            filteredRides = rides.filter({ $0.going == ridesDirectionGoing })
        }
    }
    
    func loadingFailed(withError error: NSError, checkFilteredRides: Bool = true) {
        if checkFilteredRides && filteredRides.isEmpty {
            tableView.backgroundView = errorLabel
        }
        
        NSLog("%@ failed to load rides: %@", String(describing: type(of: self)), error.localizedDescription)
        
        guard self.isVisible() else {
            return
        }
        
        if error.domain == NSURLErrorDomain {
            switch error.code {
            case NSURLErrorNotConnectedToInternet:
                CaronaeMessagesNotification.instance.showError(withText: "Sem conexão com a internet")
                return
            case NSURLErrorTimedOut,
                 NSURLErrorCannotFindHost,
                 NSURLErrorCannotConnectToHost,
                 NSURLErrorNetworkConnectionLost:
                CaronaeMessagesNotification.instance.showError(withText: "Sem conexão com o Caronaê")
                return
            default:
                break
            }
        }
        
        let errorAlertTitle = "Algo deu errado"
        let errorAlertMessage = String(format: "Não foi possível carregar as caronas. Por favor, tente novamente. (%@)", error.localizedDescription)
        CaronaeAlertController.presentOkAlert(withTitle: errorAlertTitle, message: errorAlertMessage)
    }
    
    
    // MARK: Table methods
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return filteredRides.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Ride Cell", for: indexPath) as! RideCell
        
        if historyTable {
            cell.configureHistoryCell(with: filteredRides[indexPath.row])
        } else {
            cell.configureCell(with: filteredRides[indexPath.row])
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        if !historyTable {
            let ride = filteredRides[indexPath.row]
            let rideVC = RideViewController.instance(for: ride)
            navigationController?.show(rideVC, sender: self)
        }
    }
    
    
    // MARK: IBActions
    
    @IBAction func didChangeDirection(_ sender: UISegmentedControl) {
        ridesDirectionGoing = sender.selectedSegmentIndex == 0
        updateFilteredRides()
        tableView.reloadData()
    }
    
    @IBAction func didTapFilterView(_ sender: Any) {
        performSegue(withIdentifier: "FilterRide", sender: self)
    }
    
    @IBAction func didTapClearFilterButton(_ sender: UIButton!) {
        filterIsEnabled = false
        adjustFilterView(animated: true)
    }
    
    
    // MARK: Extra views
    
    // Background view when the table is empty
    lazy var emptyTableLabel: UILabel = {
        let emptyTableLabel = UILabel(frame: CGRect(x: 0, y: 0, width: self.view.bounds.size.width, height: self.view.bounds.size.height))
        emptyTableLabel.text = self.emptyMessage ?? RideListDefault.emptyMessage
        emptyTableLabel.textColor = .gray
        emptyTableLabel.numberOfLines = 0
        emptyTableLabel.textAlignment = .center
        emptyTableLabel.font = UIFont.systemFont(ofSize: RideListDefault.messageFontSize, weight: .ultraLight)
        emptyTableLabel.sizeToFit()
        return emptyTableLabel
    }()
    
    // Background view when an error occurs
    lazy var errorLabel: UILabel = {
        let errorLabel = UILabel(frame: CGRect(x: 0, y: 0, width: self.view.bounds.size.width, height: self.view.bounds.size.height))
        errorLabel.text = RideListDefault.errorMessage
        errorLabel.textColor = .gray
        errorLabel.numberOfLines = 0
        errorLabel.textAlignment = .center
        errorLabel.font = UIFont.systemFont(ofSize: RideListDefault.messageFontSize, weight: .ultraLight)
        errorLabel.sizeToFit()
        return errorLabel
    }()

    // Background view when the table is loading
    lazy var loadingLabel: UILabel = {
        let loadingLabel = UILabel(frame: CGRect(x: 0, y: 0, width: self.view.bounds.size.width, height: self.view.bounds.size.height))
        loadingLabel.text = RideListDefault.loadingMessage
        loadingLabel.textColor = .gray
        loadingLabel.numberOfLines = 0
        loadingLabel.textAlignment = .center
        loadingLabel.font = UIFont.systemFont(ofSize: RideListDefault.messageFontSize, weight: .ultraLight)
        loadingLabel.sizeToFit()
        return loadingLabel
    }()
}
